pychromecast/socket_client.py

```diff
@@ -235,50 +255,65 @@ def initialize_connection(self):
         retry_log_fun = self.logger.error
 
         while not self.stop.is_set() and (tries is None or tries > 0):
-            try:
-                self.socket = new_socket()
-                self.socket.settimeout(self.timeout)
-                self._report_connection_status(
-                    ConnectionStatus(CONNECTION_STATUS_CONNECTING,
-                                     NetworkAddress(self.host, self.port)))
-                self.socket.connect((self.host, self.port))
-                self.socket = ssl.wrap_socket(self.socket)
-                self.connecting = False
-                self._force_recon = False
-                self._report_connection_status(
-                    ConnectionStatus(CONNECTION_STATUS_CONNECTED,
-                                     NetworkAddress(self.host, self.port)))
-                self.receiver_controller.update_status()
-                self.heartbeat_controller.ping()
-                self.heartbeat_controller.reset()
-
-                self.logger.debug("Connected!")
-                break
-            except OSError as err:
-                self.connecting = True
-                if self.stop.is_set():
-                    self.logger.error(
-                        "Failed to connect: %s. aborting due to stop signal.",
-                        err)
-                    raise ChromecastConnectionError("Failed to connect")
-
-                self._report_connection_status(
-                    ConnectionStatus(CONNECTION_STATUS_FAILED,
-                                     NetworkAddress(self.host, self.port)))
-
-                # Only sleep if we have another retry remaining
-                if tries is None or tries > 1:
-                    retry_log_fun("Failed to connect, retrying in %.1fs",
-                                  self.retry_wait)
-                    retry_log_fun = self.logger.debug
-                    time.sleep(self.retry_wait)
-
-                if tries:
-                    tries -= 1
-        else:
-            self.stop.set()
-            self.logger.error("Failed to connect. No retries.")
-            raise ChromecastConnectionError("Failed to connect")
+            for service in self.services.copy():
```

Comment:
Loop over an iterable with services, take care to copy since items may be added or removed.